### PR TITLE
[minor_change] update fix and validation

### LINKED ISSUE #957: Update function for relationship attribute "relation_l3ext_rs_subnet_to_profile" does not work in [resource_aci_l3extsubnet]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

```
"relation_l3ext_rs_subnet_to_profile": &schema.Schema{
				Type:     schema.TypeSet,
				Optional: true,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"tn_rtctrl_profile_name": {
							Type:       schema.TypeString,
							Optional:   true,
							Deprecated: "use tn_rtctrl_profile_dn instead",
						},
						"tn_rtctrl_profile_dn": {
							Type:     schema.TypeString,
							Optional: true,
						},
						"direction": {
							Type:     schema.TypeString,
							Required: true,
						},
					},
				},
			},

```

"tn_rtctrl_profile_dn" in the relationship attribute above is missing logic in the update function.
### APIC version and APIC Platform

on-prem/cloud-aws/cloud-azure/all.

### Affected Resource(s)

* aci_l3extsubnet

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

### PR DESCRIPTION
**Pull request by akinross**
fix #957

### CODE REVIEW

File: aci/resource_aci_l3extsubnet.go [hunk 1]
```diff
@@ -28,6 +28,24 @@ func resourceAciL3ExtSubnet() *schema.Resource {
 
 		SchemaVersion: 1,
 
+		CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, v interface{}) error {
+			// workaround because ValidateFunc is not allowed on TypeSet
+			if diff.HasChange("relation_l3ext_rs_subnet_to_profile") {
+				_, new := diff.GetChange("relation_l3ext_rs_subnet_to_profile")
+				// validate that a direction type ( import/export ) is not defined more than once
+				err := validateDirection(new.(*schema.Set).List())
```
[the remark this answers is not in the document]
Comment (akinross) [reply aci/resource_aci_l3extsubnet.go:36]: It is validation workaround on relationship object. Which was leading to state diff, where state has 2+ entries and the apic would only have 2.

**Review by shrsr — DISMISSED** [review]
LGTM

**Comment by sajagana** [comment]
Hi Akini,

The current code change looks good. It would be great if you can fix the below items.
1. Data Source attributes do not match with Resource attributes.
2. Remove name_alias from the attributes and use "AppendNameAliasAttrSchema" to Schema.
3. Please fix the relation_l3ext_rs_subnet_to_profile attribute import issue.
4. getRemoteL3ExtSubnet - check the DistinguishedName == "" condition.
5. Please also take care of the documentation for the data source and resources.

File: aci/resource_aci_l3extsubnet.go [hunk 2]
```diff
@@ -169,6 +226,42 @@ func setL3ExtSubnetAttributes(l3extSubnet *models.L3ExtSubnet, d *schema.Resourc
...
+		log.Printf("[DEBUG] Error while reading relation l3extRsSubnetToProfile %v", err)
+		d.Set("relation_l3ext_rs_subnet_to_profile", make([]map[string]string, 0, 1))
+		return nil, err
+	} else {
+		relParamList := make([]map[string]string, 0, 1)
+		relParams := l3extRsSubnetToProfileData.([]map[string]string)
+		for _, obj := range relParams {
+			relParamList = append(relParamList, map[string]string{
+				// obj["tnRtctrlProfileName"] is set to tDN in aci-go-client thus name is assigned to tn_rtctrl_profile_dn
+				"tn_rtctrl_profile_dn": obj["tnRtctrlProfileName"],
```
Comment (sajagana) [inline aci/resource_aci_l3extsubnet.go:241]: Just for a note, you can create a Go Client PR to fix the naming issue for tnRtctrlProfileName to tDn.

**Review by sajagana — APPROVED** [review]
LGTM!

File: aci/resource_aci_l3extsubnet.go (hunk 2 above)
Comment (akinross) [reply aci/resource_aci_l3extsubnet.go:241]: changed

**Review by anvitha-jain — DISMISSED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

File: aci/data_source_aci_l3extsubnet.go [hunk 3]
```diff
@@ -47,7 +41,34 @@ func dataSourceAciL3ExtSubnet() *schema.Resource {
...
+
+			"relation_l3ext_rs_subnet_to_profile": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"tn_rtctrl_profile_name": {
+							Type:       schema.TypeString,
+							Optional:   true,
+							Deprecated: "use tn_rtctrl_profile_dn instead",
```
Comment (lhercot) [inline aci/data_source_aci_l3extsubnet.go:53]: It is not set so you can remove the deprecation on the datasource

File: website/docs/d/l3_ext_subnet.html.markdown [hunk 4]
```diff
@@ -22,16 +31,21 @@ data "aci_l3_ext_subnet" "example" {
 
 ```
 
-## Argument Reference
+## Argument Reference ##
 
-- `external_network_instance_profile_dn` - (Required) Distinguished name of parent External Network Instance Profile object.
-- `ip` - (Required) IP address of Object l3 extension subnet.
+* `external_network_instance_profile_dn` - (Required) Distinguished name of the parent External Network Instance Profile object.
+* `ip` - (Required) IP address of the L3 Extension Subnet object.
```
Comment (lhercot) [inline website/docs/d/l3_ext_subnet.html.markdown:37]: of the External EPG Subnet

File: website/docs/d/l3_ext_subnet.html.markdown [hunk 5]
```diff
@@ -22,16 +31,21 @@ data "aci_l3_ext_subnet" "example" {
...
+* `aggregate` - (Optional) Aggregate Routes of the L3 Extension Subnet object.
+* `annotation` - (Optional) Annotation of the L3 Extension Subnet object.
+* `description` - (Optional) Description of the L3 Extension Subnet object.
+* `name_alias` - (Optional) Name alias of the L3 Extension Subnet object.
+* `scope` - (Optional) The list of domain applicable to the capability.
+* `relation_l3ext_rs_subnet_to_profile` - (Optional) Relation to Route Control Profile (class rtctrlProfile). Type: Block.
+	* `tn_rtctrl_profile_name` - **Deprecated** (Optional) Associates the External EPGs with the Route Control Profiles.
+	* `tn_rtctrl_profile_dn` - (Optional) Associates the External EPGs with the Route Control Profiles.
+	* `direction` - (Optional) Relation to configure route map for each BGP peer in the inbound and outbound directions.
+* `relation_l3ext_rs_subnet_to_rt_summ` - (Optional) Relation to class rtsumARtSummPol.
```
Comment (lhercot) [inline website/docs/d/l3_ext_subnet.html.markdown:51]: Relation to a Route Summarization Policy (class rtsumARtSummPol)

File: website/docs/r/l3_ext_subnet.html.markdown [hunk 6]
```diff
@@ -36,36 +36,32 @@ Manages ACI l3 extension subnet
...
+* `aggregate` - (Optional) Aggregate Routes of the L3 Extension Subnet object. Allowed values are "import-rtctrl", "export-rtctrl", "shared-rtctrl" and "none".
+* `annotation` - (Optional) annotation of the L3 Extension Subnet object.
+* `description` - (Optional) Description of the L3 Extension Subnet object.
+* `name_alias` - (Optional) name_alias of the L3 Extension Subnet object.
+* `scope` - (Optional) The list of domain applicable to the capability. Allowed values are "import-rtctrl", "export-rtctrl", "import-security", "shared-security" and "shared-rtctrl". Default is "import-security".
+* `relation_l3ext_rs_subnet_to_profile` - (Optional) Relation to Route Control Profile (class rtctrlProfile). Type: Block.
+	* `tn_rtctrl_profile_name` - **Deprecated** (Optional) Associates the External EPGs with the Route Control Profiles.
+	* `tn_rtctrl_profile_dn` - (Optional) Associates the External EPGs with the Route Control Profiles.
+	* `direction` - (Required) Relation to configure route map for each BGP peer in the inbound and outbound directions.
+* `relation_l3ext_rs_subnet_to_rt_summ` - (Optional) Relation to class rtsumARtSummPol.
```
Comment (lhercot) [inline website/docs/r/l3_ext_subnet.html.markdown:60]: Relation to a Route Summarization Policy (class rtsumARtSummPol)

File: website/docs/r/l3_ext_subnet.html.markdown [hunk 7]
```diff
@@ -36,36 +36,32 @@ Manages ACI l3 extension subnet
...
 
-## Importing
+## Importing ##
 
-An existing Subnet can be [imported][docs-import] into this resource via its Dn, via the following command:
+An existing L3 Extension Subnet can be [imported][docs-import] into this resource via its Dn, via the following command:
```
Comment (lhercot) [inline website/docs/r/l3_ext_subnet.html.markdown:64]: An existing External EPG Subnet

File: website/docs/d/l3_ext_subnet.html.markdown [hunk 8]
```diff
@@ -7,11 +7,20 @@ description: |-
   Data source for ACI l3 extension subnet
```
Comment (lhercot) [inline website/docs/d/l3_ext_subnet.html.markdown:7]: Data source for ACI External EPG Subnet

File: website/docs/d/l3_ext_subnet.html.markdown [hunk 9]
```diff
@@ -7,11 +7,20 @@ description: |-
...
 ---
 
-# aci_l3_ext_subnet
+# aci_l3_ext_subnet #
 
 Data source for ACI l3 extension subnet
```
Comment (lhercot) [inline website/docs/d/l3_ext_subnet.html.markdown:12]: Data source for ACI External EPG Subnet

File: website/docs/d/l3_ext_subnet.html.markdown (hunk 4 above)
Comment (lhercot) [reply website/docs/d/l3_ext_subnet.html.markdown:37]: change for all others too.

File: website/docs/r/l3_ext_subnet.html.markdown [hunk 10]
```diff
@@ -7,7 +7,7 @@ description: |-
   Manages ACI l3 extension subnet
```
Comment (lhercot) [inline website/docs/r/l3_ext_subnet.html.markdown:7]: Manages ACI External EPG Subnet

File: website/docs/r/l3_ext_subnet.html.markdown [hunk 11]
```diff
@@ -7,7 +7,7 @@ description: |-
...
 ---
 
-# aci_l3_ext_subnet
+# aci_l3_ext_subnet #
 
 Manages ACI l3 extension subnet
```
Comment (lhercot) [inline website/docs/r/l3_ext_subnet.html.markdown:12]: Manages ACI External EPG Subnet

File: website/docs/r/l3_ext_subnet.html.markdown [hunk 12]
```diff
@@ -36,36 +36,32 @@ Manages ACI l3 extension subnet
...
-
-* `relation_l3ext_rs_subnet_to_profile` - (Optional) Relation to class rtctrlProfile. Cardinality - N_TO_M. Type - Set of Map.
-* `relation_l3ext_rs_subnet_to_rt_summ` - (Optional) Relation to class rtsumARtSummPol. Cardinality - N_TO_ONE. Type - String.
-
-## Attribute Reference
-
-The only attribute that this resource exports is the `id`, which is set to the
-Dn of the l3 extension subnet.
+* `external_network_instance_profile_dn` - (Required) Distinguished name of the parent External Network Instance Profile object.
+* `ip` - (Required) ip of the L3 Extension Subnet object.
```
Comment (lhercot) [inline website/docs/r/l3_ext_subnet.html.markdown:50]: of the External EPG Subnet
Update all other references too.

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/data_source_aci_l3extsubnet.go (hunk 3 above)
Comment (akinross) [reply aci/data_source_aci_l3extsubnet.go:53]: ok

File: website/docs/d/l3_ext_subnet.html.markdown (hunk 4 above)
Comment (akinross) [reply website/docs/d/l3_ext_subnet.html.markdown:37]: ok

File: website/docs/d/l3_ext_subnet.html.markdown (hunk 5 above)
Comment (akinross) [reply website/docs/d/l3_ext_subnet.html.markdown:51]: ok

File: website/docs/r/l3_ext_subnet.html.markdown (hunk 6 above)
Comment (akinross) [reply website/docs/r/l3_ext_subnet.html.markdown:60]: ok

File: website/docs/r/l3_ext_subnet.html.markdown (hunk 7 above)
Comment (akinross) [reply website/docs/r/l3_ext_subnet.html.markdown:64]: ok

File: website/docs/d/l3_ext_subnet.html.markdown (hunk 8 above)
Comment (akinross) [reply website/docs/d/l3_ext_subnet.html.markdown:7]: ok

File: website/docs/d/l3_ext_subnet.html.markdown (hunk 9 above)
Comment (akinross) [reply website/docs/d/l3_ext_subnet.html.markdown:12]: ok

File: website/docs/r/l3_ext_subnet.html.markdown (hunk 11 above)
Comment (akinross) [reply website/docs/r/l3_ext_subnet.html.markdown:12]: ok

File: website/docs/r/l3_ext_subnet.html.markdown (hunk 12 above)
Comment (akinross) [reply website/docs/r/l3_ext_subnet.html.markdown:50]: ok

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**